### ISSUE TITLE
Question: What is the behavior of `read_timeout` in the HTTP client?

### ISSUE THREAD

**Issue by QnJ1c2kNCg** [smithy-lang/smithy-rs#3145]
Hello, I would like to better understand how the `read_timeout` works with the HTTP client.

1. Could you share a code link to _where_ this timeout is used? (I could not find it)
2. What is considered first byte? Could the server send some bytes that are not part of the response and not timeout?

More context: In our application, we've set this `read_timeout` to a relatively small value i.e 3 seconds, and everything worked fine (no timeout errors observed). But we had a suspicion that some HTTP call took longer than 3 seconds and that the client seemed to _hang_.

We then wrapped all of our HTTP calls with a `tokio::time::timeout` that is **much** larger i.e 10 minutes, and noticed that that 10 minutes timeout was getting hit.

What would be some examples where `read_timeout` doesn't timeout, but the HTTP client call completion `tokio::time::timeout`s?

AFAICT, our server application sends small responses (no big bodies) and doesn't send anything explicitly to keep the connections open.

Thanks!

**Comment by jdisanti** [comment]
The read and connect timeouts are implemented as part of the Smithy hyper connector. The connect timeout wraps the TLS connector, and the read timeout wraps the hyper client. Thus, the read timeout is from when the request is sent until the hyper client returns a response, which should occur once the response headers have come in.

If you're observing hanging with these timeouts set, it is likely due to the response body, which doesn't have a timeout. If the server stops sending data mid-stream, then that can cause an indefinite hang on the client side. A [body minimum throughput timeout](https://github.com/awslabs/smithy-rs/pull/3068) has been implemented and should be available for use in the next release for exactly this issue.

Code for the read/connect timeouts: https://github.com/awslabs/smithy-rs/blob/main/rust-runtime/aws-smithy-runtime/src/client/http/hyper_014.rs

**Comment by jdisanti** [comment]
Actually, it looks like the implementation was already released in 0.57.x: https://docs.rs/aws-smithy-runtime/latest/aws_smithy_runtime/client/http/body/minimum_throughput/struct.MinimumThroughputBody.html

However, I don't think there are currently any conveniences that wire it up right now. @Velfi, is this thing usable yet?

**Comment by Velfi** [comment]
It's technically usable, but we still have a lot of UX design to do before we make it default. If anyone would like to try it out, you can do so like this:

```rust
use std::time::Duration;
use aws_sdk_s3::primitives::SdkBody;
use aws_smithy_runtime::client::http::body::minimum_throughput::MinimumThroughputBody;

#[tokio::main]
async fn main() {
    let sdk_config = aws_config::load_from_env().await;
    let client = aws_sdk_s3::Client::new(&sdk_config);
    
    let res = client.get_object()
    .bucket("zhessler-test-bucket")
    .key("100MBTest.txt")
    .send()
    .await.expect("request succeeds");

    let time_source = sdk_config.time_source().clone().expect("a default time source is set");
    let async_sleep = sdk_config.sleep_impl().clone().expect("a default async sleep implementation is set");
    let body = res.body.map(move |body| {
        let minimum_throughput = (1000u64, Duration::from_secs(1)); // 1 KB/s
        let time_source = time_source.clone();
        let async_sleep = async_sleep.clone();
        SdkBody::from_body_0_4(
            MinimumThroughputBody::new(time_source, async_sleep, body, minimum_throughput),
        )
    });
    // Stream the rest of the body into memory. If the body streams too slowly,
    // then the throughput timeout will trigger.
    let _body = body.collect().await.expect("body streams without error");
}
```

Let me know if you have any questions.
